### ISSUE TITLE
Ktorm: issues & suggestions

### ISSUE THREAD

**Issue by halimpuckjava** [vincentlauvlwj/Ktorm#26]
hi

1.
i use postgresql dialect (not mysql) but when invoke method insertorupadate i see this exception: 
```
Unsupported expression type: class me.liuwj.ktorm.support.mysql.InsertOrUpdateExpression
```

2. how to release (disconnect) from database when finish (i use Database.connect { ..... }) i not found any method in Database class to disconnect from db

3. add interceptor when database disconnect 

4. add ktorm support for : sqlite, firebird, mssql server, mongo

**Comment by vincentlauvlwj** [comment]
1. The `insertOrUpdate` function is provided in ktorm-support-mysql module. It's designed to support MySQL's `on duplicate key update` syntax, not for PostgreSQL. So you can not use it if you are using PostgreSQL dialect. BTW, PostgreSQL has its own `on conflict ... do update` syntax, but it's not supported by Ktorm yet (will fix in the future). 

2. We don't need to release a `Database` instance because Ktorm closes connections automatically. `Database` doesn't represent real connections. It just holds some metadata and configurations used to obtain connections. Ktorm obtains connections from a `Database` instance only when it's really needed (such as executing a SQL), then close them immediately after they are not useful anymore. More details can be seen in our documentation: [Connect to Databases](https://ktorm.liuwj.me/en/connect-to-databases.html).

3. Because we don't need to close a `Database` instance, so this makes no sense. 

4. Ktorm is built based on JDBC and its core module generates standard SQLs, so any database that provides a JDBC driver should be supported naturally. Ktorm's dialect modules just provide some database specific syntax and SQL functions. Without those dialect modules, we can still access any relational database via the standard SQLs generated by ktorm-core.

**Comment by halimpuckjava** [comment]
thanks,

**Comment by vincentlauvlwj** [comment]
Ktorm is designed for relational databases.

As for NoSQL, all I can say is _maybe_. It possibly works if there is a JDBC driver, but it's not well tested.

**Comment by vincentlauvlwj** [comment]
I just released a new version v2.4 hours ago: https://github.com/vincentlauvlwj/Ktorm/releases/tag/v2.4

The new version provides an `insertOrUpdate` function for PostgreSQL, located in package `me.liuwj.ktorm.support.postgresql`.

**Comment by halimpuckjava** [comment]
great
